newton/sim/ik.py

```diff
@@ -0,0 +1,1597 @@
+            for sync_event in self.sync_events:
+                main.wait_event(sync_event)
+        else:
+            for obj, offset in zip(self.objectives, self.residual_offsets):
+                fn(obj, offset, *extra)
+
+    def solve(self, iterations=10, step_size=1.0):
+        """
+        Run the Levenberg-Marquardt loop.
+
+        Parameters
+        ----------
+        iterations : int, default 10
+            Maximum number of outer iterations.
+        step_size : float, default 1.0
+            Multiplicative scale on deltaq before applying the proposal.
+
+        Side-effects
+        ------------
+        Updates `self.joint_q` in-place with the converged coordinates.
+        """
+        self.lambda_values.fill_(self.lambda_initial)
+        for i in range(iterations):
+            self._step(step_size=step_size, iteration=i)
+
+    def compute_residuals(self, joint_q=None, output_residuals=None):
+        joint_q = joint_q or self.joint_q
+        output_residuals = output_residuals or self.residuals
+
+        if self.jacobian_mode in [JacobianMode.AUTODIFF, JacobianMode.MIXED]:
+            _eval_fk_batched(self.model, joint_q, self.joint_qd, self.body_q, self.body_qd)
+        else:
+            self._fk_two_pass(self.model, joint_q, self.body_q, self.X_local, self.n_problems)
+
+        output_residuals.zero_()
+
+        def _do(obj, off, body_q, joint_q, model, output_residuals):
+            obj.compute_residuals(body_q, joint_q, model, output_residuals, off)
+
+        self._parallel_for_objectives(_do, self.body_q, joint_q, self.model, output_residuals)
+
+        return output_residuals
+
+    def compute_jacobian(self):
+        self.jacobian.zero_()
+
+        if self.jacobian_mode == JacobianMode.AUTODIFF:
+            self.tape.reset()
+            with self.tape:
+                residuals_2d = self.compute_residuals(self.joint_q)
+                current_residuals_wp = residuals_2d.flatten()
+
+            self.tape.outputs = [current_residuals_wp]
+
+            for obj, offset in zip(self.objectives, self.residual_offsets):
+                obj.compute_jacobian_autodiff(self.tape, self.model, self.jacobian, offset, self.joint_q)
+                self.tape.zero()
+
+        elif self.jacobian_mode == JacobianMode.ANALYTIC:
+            n_joints = self.model.joint_count
+
+            wp.launch(
+                self._compute_motion_subspace_2d,
+                dim=self.n_problems * n_joints,
+                inputs=[
+                    self.n_problems,
+                    self.model.joint_type,
+                    self.model.joint_parent,
+                    self.model.joint_qd_start,
+                    self.joint_qd,
+                    self.model.joint_axis,
+                    self.model.joint_dof_dim,
+                    self.body_q,
+                    self.model.joint_X_p,
+                ],
+                outputs=[
+                    self.joint_S_s,
+                ],
```

Comment:
Make sure the `device` argument is set correctly (probably `self.model.device`).